src/poly/evaluator.rs

```diff
@@ -0,0 +1,542 @@
+    }
+}
+
+impl<E: Clone, F: Field, B: Basis> Mul<F> for &Ast<E, F, B> {
+    type Output = Ast<E, F, B>;
+
+    fn mul(self, other: F) -> Self::Output {
+        Ast::Scale(Box::new(self.clone()), other)
+    }
+}
+
+impl<E: Clone, F: Field> MulAssign for Ast<E, F, ExtendedLagrangeCoeff> {
+    fn mul_assign(&mut self, rhs: Self) {
+        *self = self.clone().mul(rhs)
+    }
+}
+
+/// Operations which can be performed over a given basis.
+pub(crate) trait BasisOps: Basis {
+    fn empty_poly<F: FieldExt>(domain: &EvaluationDomain<F>) -> Polynomial<F, Self>;
+    fn constant_term<F: FieldExt>(chunk_size: usize, chunk_index: usize, scalar: F) -> Vec<F>;
+    fn linear_term<F: FieldExt>(
+        domain: &EvaluationDomain<F>,
+        chunk_size: usize,
+        chunk_index: usize,
+        scalar: F,
+    ) -> Vec<F>;
+    fn rotate<F: FieldExt>(
+        domain: &EvaluationDomain<F>,
+        poly: &Polynomial<F, Self>,
+        rotation: Rotation,
+    ) -> Polynomial<F, Self>;
+}
+
+impl BasisOps for Coeff {
+    fn empty_poly<F: FieldExt>(domain: &EvaluationDomain<F>) -> Polynomial<F, Self> {
+        domain.empty_coeff()
+    }
+
+    fn constant_term<F: FieldExt>(chunk_size: usize, chunk_index: usize, scalar: F) -> Vec<F> {
+        let mut chunk = vec![F::zero(); chunk_size];
+        if chunk_index == 0 {
+            chunk[0] = scalar;
+        }
+        chunk
+    }
+
+    fn linear_term<F: FieldExt>(
+        _: &EvaluationDomain<F>,
+        chunk_size: usize,
+        chunk_index: usize,
+        scalar: F,
+    ) -> Vec<F> {
+        let mut chunk = vec![F::zero(); chunk_size];
+        // If the chunk size is 1 (e.g. if we have a small k and many threads), then the
+        // linear coefficient is the second chunk. Otherwise, the chunk size is greater
+        // than one, and the linear coefficient is the second element of the first chunk.
+        if chunk_size == 1 && chunk_index == 1 {
+            chunk[0] = scalar;
+        } else if chunk_index == 0 {
+            chunk[1] = scalar;
+        }
+        chunk
+    }
+
+    fn rotate<F: FieldExt>(
+        _: &EvaluationDomain<F>,
+        _: &Polynomial<F, Self>,
+        _: Rotation,
+    ) -> Polynomial<F, Self> {
+        panic!("Can't rotate polynomials in the standard basis")
+    }
+}
+
+impl BasisOps for LagrangeCoeff {
+    fn empty_poly<F: FieldExt>(domain: &EvaluationDomain<F>) -> Polynomial<F, Self> {
+        domain.empty_lagrange()
+    }
+
+    fn constant_term<F: FieldExt>(chunk_size: usize, _: usize, scalar: F) -> Vec<F> {
+        vec![scalar; chunk_size]
+    }
+
+    fn linear_term<F: FieldExt>(
+        domain: &EvaluationDomain<F>,
+        chunk_size: usize,
+        chunk_index: usize,
+        scalar: F,
+    ) -> Vec<F> {
+        // Take every power of omega within the chunk, and multiply by scalar.
+        let omega = domain.get_omega();
+        let start = chunk_size * chunk_index;
+        (0..chunk_size)
```

Comment:
Does this work correctly for the final (potentially smaller) chunk?

Reply:
Yes, as long as the AST isn't simply `Ast::LinearTerm`. In any non-trivial AST, this term will be combined via either `Ast::Add` or `Ast::Mul`; if the final chunk is short, then the `.zip()` in either case's impl will ensure we only process up to the end of the chunk, dropping the remainder of the `LinearTerm`.

Reply:
Ok; to avoid unexpected surprises (e.g. if the code is reused) we should probably make it work for the trivial case as well.